### PR TITLE
Move electrical profile set id from timetable to scenario

### PR DESCRIPTION
**Pull request by Castavo**
> [!Warning]
> For editoast please review only the first commit

The timetable model gets a ton of changes in editoast because the `ModelV2` derive does not handle struct with only primary keys.
Indeed, the derive creates a `TimetableV2Changeset` struct, with no fields, and a struct with no field can not derive `AsChangeset` and `Insertable` from `diesel`.
With @leovalais we checked the following solutions:
* implementing the diesel traits manually on `TimetableV2Changeset`: impossible due to the way they're made
* update the derive so in the case there are only primary keys in the struct, we do not rely on a `TimetableV2Changeset` struct: this would complexify the derive a lot for a single and simple use, we deemed it overkill.
* updating the derive so that you could pick which parts of the CRUD you wish to be automatically derived: most desirable but very havy in resources, should come at some point
* implementing the `ModelV2` traits manually: the solution we selected.

I removed the timetable listing endpoint because it was unused, because the `List` trait required the struct to be a `Model` (which it isn't anymore), and because a list of only ids did not seem very exploitable...

### CODE REVIEW

**Review by flomonster — COMMENTED** [review]
Great PR. Not tested yet (waiting to be ready for review)

**Review by younesschrifi — COMMENTED** [review]
LGTM For editoast part :)

**Review by Wadjetz — APPROVED** [review]
LGTM

File: editoast/src/views/v2/timetable.rs [hunk 1]
```diff
@@ -111,8 +97,7 @@ impl From<TimetableWithTrains> for TimetableDetailedResult {
     fn from(val: TimetableWithTrains) -> Self {
         Self {
             timetable: TimetableResult {
-                id: val.id,
-                electrical_profile_set_id: val.electrical_profile_set_id,
+                timetable_id: val.id,
```
Comment (flomonster) [inline editoast/src/views/v2/timetable.rs]: We could adapt the `TimetableDetailedResult` to:

```rust
/// Creation result for a Timetable
#[derive(Debug, Default, Serialize, Deserialize, Derivative, ToSchema)]
#[cfg_attr(test, derive(PartialEq))]
struct TimetableDetailedResult {
    pub timetable_id: i64,
    pub train_ids: Vec<i64>,
}
```

**Review by flomonster — APPROVED** [review]
LGTM, (tested + reviewed)

**Review by leovalais — APPROVED** [review]
LGTM, not tested. An issue tracking the problem with ModelV2 is opened here: https://github.com/OpenRailAssociation/osrd/issues/8111

**Review by shenriotpro — APPROVED** [review]
OK for tests/

File: front/src/reducers/osrdconf/osrdConfCommon/selectors.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ const buildCommonConfSelectors = <ConfState extends OsrdConfState>(
     getConstraintDistribution: makeOsrdConfSelector('constraintDistribution'),
     getName: makeOsrdConfSelector('name'),
     getTrainCount: makeOsrdConfSelector('trainCount'),
+    // TODO : là c'est le get de la valeur des trucs
```
Comment (SharglutDev) [inline front/src/reducers/osrdconf/osrdConfCommon/selectors.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/index.ts [hunk 3]
```diff
@@ -73,12 +73,14 @@ interface CommonConfReducers<S extends OsrdConfState> extends InfraStateReducers
   ['updateTrainStep']: CaseReducer<S, PayloadAction<S['trainStep']>>;
   ['updateAllowances']: CaseReducer<S, PayloadAction<S['allowances']>>;
   ['toggleUsingElectricalProfiles']: CaseReducer<S>;
+  // TODO: update ep set id
```
Comment (SharglutDev) [inline front/src/reducers/osrdconf/osrdConfCommon/index.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/__tests__/utils.ts [hunk 4]
```diff
@@ -275,6 +282,8 @@ const testCommonConfReducers = (slice: OperationalStudiesConfSlice | StdcmConfSl
     });
   });
 
+  // TODO: add test for setting then reading. => pour faire tourner: yarn test.
+
```
Comment (SharglutDev) [inline front/src/reducers/osrdconf/osrdConfCommon/__tests__/utils.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/operationalStudiesConf/simulationConfReducers.spec.ts [hunk 5]
```diff
@@ -21,3 +21,5 @@ describe('simulationConfReducer', () => {
 
   testCommonConfReducers(operationalStudiesConfSlice);
 });
+
+// TODO: c'est ce chemin uq'il faut donner a yarn test
```
Comment (SharglutDev) [inline front/src/reducers/osrdconf/operationalStudiesConf/simulationConfReducers.spec.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/index.ts
Comment (SharglutDev) [inline front/src/reducers/osrdconf/osrdConfCommon/index.ts:1]: Could you add `electricalProfileSetID: undefined,` in `defaultCommonConf` const ? (So we can remember that it's there even if it's initialized with undefined)

File: front/src/modules/scenario/components/ScenarioCard.tsx [hunk 6]
```diff
@@ -14,16 +14,16 @@ import { updateSelectedProjection } from 'reducers/osrdsimulation/actions';
...
 import { dateTimeFormatting } from 'utils/date';
 
-type StudyCardProps = {
+type ScenarioCardProps = {
   setFilterChips: (filterChips: string) => void;
   scenario: ScenarioWithCountTrains;
 };
 
-export default function StudyCard({ setFilterChips, scenario }: StudyCardProps) {
+export default function ScenarioCard({ setFilterChips, scenario }: ScenarioCardProps) {
```
Comment (clarani) [inline front/src/modules/scenario/components/ScenarioCard.tsx:22]: Nice catch :)

File: front/src/modules/scenario/components/ScenarioCard.tsx [hunk 7]
```diff
@@ -14,16 +14,16 @@ import { updateSelectedProjection } from 'reducers/osrdsimulation/actions';
...
   const { t } = useTranslation('operationalStudies/study');
   const navigate = useNavigate();
   const dispatch = useAppDispatch();
-  const { updateScenarioID } = useOsrdConfActions();
+  const { updateScenarioID } = useOsrdConfActions(); // Update après
```
Comment (clarani) [inline front/src/modules/scenario/components/ScenarioCard.tsx]: You forgot to remove this comment
```suggestion
  const { updateScenarioID } = useOsrdConfActions();
```

File: front/src/reducers/osrdconf/operationalStudiesConf/simulationConfReducers.spec.ts (hunk 5 above)
Comment (clarani) [inline front/src/reducers/osrdconf/operationalStudiesConf/simulationConfReducers.spec.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/__tests__/utils.ts [hunk 8]
```diff
@@ -275,6 +282,8 @@ const testCommonConfReducers = (slice: OperationalStudiesConfSlice | StdcmConfSl
     });
   });
 
+  // TODO: add test for setting then reading. => pour faire tourner: yarn test.
```
Comment (clarani) [inline front/src/reducers/osrdconf/osrdConfCommon/__tests__/utils.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/index.ts (hunk 3 above)
Comment (clarani) [inline front/src/reducers/osrdconf/osrdConfCommon/index.ts]:
```suggestion
```

File: front/src/reducers/osrdconf/osrdConfCommon/selectors.ts (hunk 2 above)
Comment (clarani) [inline front/src/reducers/osrdconf/osrdConfCommon/selectors.ts]:
```suggestion
```

File: front/tests/007-op-rollingstock-tab.spec.ts [hunk 9]
```diff
@@ -7,22 +7,22 @@ import type {
...
-import { getProject, getStudy, getRollingStock, postApiRequest, getInfra } from './utils/index';
+import { getInfra, getProject, getRollingStock, getStudy, postApiRequest } from './utils/index';
 
 let smallInfra: Infra;
 let project: Project;
 let study: Study;
 let scenario: Scenario;
 let rollingStock: RollingStock;
-let timetable: Timetable;
+let timetable_result: TimetableResult;
```
Comment (clarani) [inline front/tests/007-op-rollingstock-tab.spec.ts]: We don't use snake case in the front :) timetable or timetableResult is better
```suggestion
let timetable: TimetableResult;
```

File: front/tests/010-op-route-tab.spec.ts [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
 import ScenarioPage from './pages/scenario-page-model';
-import { getProject, getStudy, postApiRequest, getInfra } from './utils/index';
+import { getInfra, getProject, getStudy, postApiRequest } from './utils/index';
 
 let smallInfra: Infra;
 let project: Project;
 let study: Study;
 let scenario: Scenario;
-let timetable: Timetable;
+let timetable_result: TimetableResult;
```
Comment (clarani) [inline front/tests/010-op-route-tab.spec.ts]: same here

File: front/src/reducers/osrdconf/types.ts [hunk 11]
```diff
@@ -29,6 +29,7 @@ export interface OsrdConfState extends InfraState {
   scenarioID?: number;
   pathfindingID?: number;
   timetableID?: number;
+  electricalProfileSetID?: number | null;
```
Comment (clarani) [inline front/src/reducers/osrdconf/types.ts]: Do you think it would be possible to have only a number or undefined ? You might need to add a test when you dispatch an action updateElectricalProfileSetId.
Also, I would like to get rid of the ID in maj, can you use Id in classic camel case ? 
```suggestion
  electricalProfileSetId?: number;
```

File: front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx [hunk 12]
```diff
@@ -96,12 +96,13 @@ const ScenarioV2 = () => {
...
   }, [isScenarioError, errorScenario]);
 
-  const { updateInfraID, updateTimetableID } = useOsrdConfActions();
+  const { updateInfraID, updateTimetableID, updateElectricalProfileSetID } = useOsrdConfActions();
 
   useEffect(() => {
     if (scenario) {
       dispatch(updateTimetableID(scenario.timetable_id));
       dispatch(updateInfraID(scenario.infra_id));
+      dispatch(updateElectricalProfileSetID(scenario.electrical_profile_set_id));
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx]: - use Id instead of ID in the name of the action
- add a small test to dispatch undefined rather than null (I'm not sure this suggestion works, but you can try :) otherwise, a classic ternary test with `... ? ... : ...` will do the job)
```suggestion
      dispatch(updateElectricalProfileSetId(scenario.electrical_profile_set_id || undefined));
```

File: front/src/modules/scenario/components/ScenarioCard.tsx (hunk 7 above)
Comment (SharglutDev) [inline front/src/modules/scenario/components/ScenarioCard.tsx]:
```suggestion
  const { updateScenarioID } = useOsrdConfActions();
```

Comment (Castavo) [reply front/src/modules/scenario/components/ScenarioCard.tsx]: My bad, I was tired

File: front/tests/007-op-rollingstock-tab.spec.ts (hunk 9 above)
Comment (Castavo) [reply front/tests/007-op-rollingstock-tab.spec.ts]: Ah sorry, I slipped

File: front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx [hunk 13]
```diff
@@ -113,7 +114,11 @@ const ScenarioV2 = () => {
   );
 
   const { data: conflicts } = osrdEditoastApi.endpoints.getV2TimetableByIdConflicts.useQuery(
-    { id: timetable?.id as number, infraId: infraId! },
+    {
+      id: timetable?.timetable_id as number,
+      infraId: infraId!,
+      electricalProfileSetId: scenario?.electrical_profile_set_id,
+    },
     { skip: !timetable || !infraId }
```
Comment (SharglutDev) [inline front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx]: To harmonize with other params, we should either : 
- get electricalProfilSetId with a useSelector and use it in the endpoint and add `!electricalProfilSetId` in `skip`

```suggestion
      const { data: conflicts } = osrdEditoastApi.endpoints.getV2TimetableByIdConflicts.useQuery(
    {
      id: timetable?.timetable_id as number,
      infraId: infraId!,
      electricalProfileSetId: electricalProfileSetId,
    },
    { skip: !timetable || !infraId || !electricalProfileSetId }
  );
```

- chose to get all these infos from the scenario as it carries all we need

```suggestion
    const { data: conflicts } = osrdEditoastApi.endpoints.getV2TimetableByIdConflicts.useQuery(
    {
      id: scenario?.timetable_id as number,
      infraId: scenario?.infra_id as number,
      electricalProfileSetId: scenario?.electrical_profile_set_id,
    },
    { skip: !scenario }
  );
```

File: front/tests/010-op-route-tab.spec.ts (hunk 10 above)
Comment (SharglutDev) [inline front/tests/010-op-route-tab.spec.ts]: When possible, we don't use snake case in front :)

```suggestion
let timetableResult: TimetableResult;
```

File: front/tests/007-op-rollingstock-tab.spec.ts (hunk 9 above)
Comment (SharglutDev) [inline front/tests/007-op-rollingstock-tab.spec.ts]: Same here

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for this nice PR, left some comments (mostly for syntax and front good practices)

File: front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx (hunk 13 above)
Comment (Castavo) [reply front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx]: I picked your second option

File: front/src/common/api/generatedEditoastApi.ts [hunk 14]
```diff
@@ -888,7 +869,10 @@ const injectedRtkApi = api
       >({
         query: (queryArg) => ({
           url: `/v2/timetable/${queryArg.id}/conflicts`,
-          params: { infra_id: queryArg.infraId },
+          params: {
+            infra_id: queryArg.infraId,
+            electrical_profile_set_id: queryArg.electricalProfileSetId,
+          },
```
Comment (SharglutDev) [inline front/src/common/api/generatedEditoastApi.ts:875]: When creating a scenario without electrical profile, I have this error : 

Capture d’écran 2024-07-30 à 17 55 11

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for the changes, one issue left

File: front/src/common/api/generatedEditoastApi.ts (hunk 14 above)
Comment (Castavo) [reply front/src/common/api/generatedEditoastApi.ts:875]: Oh no

Comment (SharglutDev) [reply front/src/common/api/generatedEditoastApi.ts:875]: Ok my bad, I didn't rebuild editoast after your last changes

**Comment by emersion** [comment]
Note to self: we're now passing the electrical set ID as query param in requests because a single timetable might be used from multiple scenarii.

File: editoast/src/views/v2/timetable.rs [hunk 15]
```diff
@@ -151,95 +132,22 @@ async fn get(
...
-        .into_selection_settings();
-    let (timetables, stats) =
-        Timetable::list_paginated(db_pool.get().await?.deref_mut(), settings).await?;
-    Ok(Json(ListTimetablesResponse {
-        stats,
-        results: timetables.into_iter().map_into().collect(),
-    }))
-}
-
-/// Create a timetable
```
Comment (emersion) [inline editoast/src/views/v2/timetable.rs:186]: Nit: can we keep this code comment here? Without this the OpenAPI endpoint is missing a description.

File: editoast/src/views/v2/timetable.rs [hunk 16]
```diff
@@ -74,46 +68,33 @@ enum TimetableError {
...
-}
-
-/// Creation form for a Timetable
+/// Creation result for a Timetable
 #[derive(Debug, Default, Serialize, Deserialize, Derivative, ToSchema)]
+#[cfg_attr(test, derive(PartialEq))]
 struct TimetableResult {
-    pub id: i64,
-    pub electrical_profile_set_id: Option<i64>,
+    pub timetable_id: i64,
```
Comment (emersion) [inline editoast/src/views/v2/timetable.rs:75]: Completely inconsequential nit: any motivation for the `id` → `timetable_id` rename? Since the struct already has "Timetable" in the name `id` may be descriptive enough already?

Comment (Castavo) [reply editoast/src/views/v2/timetable.rs:75]: The reasoning is that the endpoint's response is pure json so "id" isn't extremely descriptive. You can deduce it's a timetable_id from the name of the endpoint, but we might as well be extra clear ?

File: editoast/src/tables.rs [hunk 17]
```diff
@@ -593,6 +594,21 @@ diesel::table! {
...
+    stdcm_search_environment (id) {
+        id -> Int8,
+        infra_id -> Int8,
+        electrical_profile_set_id -> Nullable<Int8>,
+        work_schedule_group_id -> Nullable<Int8>,
+        timetable_id -> Int8,
+        search_window_begin -> Timestamptz,
+        search_window_end -> Timestamptz,
+    }
+}
```
Comment (emersion) [inline editoast/src/tables.rs]: Seems like this is unused/unrelated :P

Comment (Castavo) [reply editoast/src/tables.rs]: Indeed, my bad

**Review by emersion — COMMENTED** [review]
Apart from these minor comments, looks good!

File: editoast/src/views/v2/timetable.rs (hunk 16 above)
Comment (emersion) [reply editoast/src/views/v2/timetable.rs:75]: Fine by me!

File: editoast/src/views/v2/timetable.rs (hunk 1 above)
Comment (Castavo) [reply editoast/src/views/v2/timetable.rs]: Done

**Review by SharglutDev — APPROVED** [review]
Lgtm and tested, great job :)

File: front/src/reducers/osrdconf/types.ts (hunk 11 above)
Comment (SharglutDev) [reply front/src/reducers/osrdconf/types.ts]: Null has been removed so it's all good

File: front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx (hunk 12 above)
Comment (SharglutDev) [reply front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx]: Null has been removed so it's all good
